pkg/csi/service/vanilla/controller.go

```diff
@@ -1243,14 +1244,75 @@ func (c *controller) CreateSnapshot(ctx context.Context, req *csi.CreateSnapshot
 				"cannot snapshot migrated vSphere volume. :%q", volumeID)
 		}
 
-		// Query Capacity in MB for block volume snapshot
-		snapshotSizeInMB, err := utils.QueryCapacityForBlockVolumeSnapshotUtil(
+		// Query capacity in MB and datastore url for block volume snapshot
+		snapshotSizeInMB, datastoreUrl, err := utils.QueryCapacityAndDsUrlForVolumeUtil(
 			ctx, c.manager.VolumeManager, volumeID, common.BlockVolumeType)
 		if err != nil {
 			return nil, err
 		}
 		volumeType = prometheus.PrometheusBlockVolumeType
 
+		// Check if snapshots number of this volume reaches the granular limit on VSAN/VVOL
+		maxSnapshotsPerBlockVolume := c.manager.CnsConfig.Snapshot.GlobalMaxSnapshotsPerBlockVolume
+		log.Infof("The limit of the maximum number of snapshots per block volume is "+
+			"set to the global maximum (%v) by default.", maxSnapshotsPerBlockVolume)
+		if c.manager.CnsConfig.Snapshot.GranularMaxSnapshotsPerBlockVolumeInVSAN != 0 ||
+			c.manager.CnsConfig.Snapshot.GranularMaxSnapshotsPerBlockVolumeInVVOL != 0 {
```

Comment:
If you used ">0" comparison you can also implicitly validate that values are always a positive integer.